spec/abci++/v2.md

```diff
@@ -0,0 +1,180 @@
+# Tendermint v2 Markdown pseudocode
+
+This adds a single-threaded implementation of ABCI++,
+with no optimization for splitting out verifying the header and verifying the proposal.
+
+### Initialization
+
+```go
+h_p ← 0
+round_p ← 0
+step_p is one of {propose, prevote, precommit}
+decision_p ← nil
+lockedValue_p ← nil
+validValue_p ← nil
+validRound_p ← -1
+```
+
+### StartRound(round)
+
+```go
+function startRound(round) {
+    round_p ← round
+    step_p ← propose
+    if proposer(h_p, round_p) = p {
+        if validValue_p != nil {
+            proposal ← validValue_p
+        } else {
+            txdata ← mempool.GetBlock()
+            // getBlockProposal fills in header
+            unpreparedProposal ← getUnpreparedBlockProposal(txdata)
+            proposal ← ABCI.PrepareProposal(unpreparedProposal)
+        }
+        broadcast ⟨PROPOSAL, h_p, round_p, proposal, validRound_p⟩
+    } else {
+        schedule OnTimeoutPropose(h_p,round_p) to be executed after timeoutPropose(round_p)
+    }
+}
+```
+
+### ReceiveProposal
+
+In the case where the local node is not locked on any round, the following is ran:
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, −1) from proposer(h_p, round_p) while step_p = propose do {
+    if valid(v) ∧ ABCI.ProcessProposal(h_p, v).accept ∧ (lockedRound_p = −1 ∨ lockedValue_p = v) {
+        broadcast ⟨PREVOTE, h_p, round_p, id(v,t)⟩ 
+    } else {
+        broadcast ⟨PREVOTE, h_p, round_p, nil⟩ 
+        // Include any slashing evidence that may be sent in the process proposal response
+        for evidence in ABCI.ProcessProposal(h_p, v).evidence_list {
+            broadcast ⟨EVIDENCE, evidence⟩ 
+        }
+    }
+    step_p ← prevote
+}
+```
+
+In the case where the node is locked on a round, the following is ran:
+
+```go
+upon ⟨PROPOSAL, h_p, round_p, v, vr⟩
+  from proposer(h_p, round_p)
+  AND 2f + 1 ⟨PREVOTE, h_p, vr, id(v)⟩ 
+  while step_p = propose ∧ (vr ≥ 0 ∧ vr < round_p) do {
+    if valid(v) ∧ ABCI.ProcessProposal(h_p, v).accept ∧ (lockedRound_p ≤ vr ∨ lockedValue_p = v) {
+        broadcast ⟨PREVOTE, h_p, round_p, id(v)⟩
+    } else {
+        broadcast ⟨PREVOTE, h_p, round_p, nil⟩
+        // Include any slashing evidence that may be sent in the process proposal response
+        for evidence in ABCI.ProcessProposal(h_p, v).evidence_list {
+            broadcast ⟨EVIDENCE, evidence⟩ 
+        }
+    }
+    step_p ← prevote
+}
+```
+
+### Prevote timeout
+
+Upon receiving 2f + 1 prevotes, setup a timeout.
+
+```go
+upon 2f + 1 ⟨PREVOTE, h_p, vr, -1⟩ 
+  with step_p = prevote for the first time, do {
+    schedule OnTimeoutPrevote(h_p, round_p) to be executed after timeoutPrevote(round_p)
+}
+```
+
+with OnTimeoutPrevote defined as:
+
+```go
+function OnTimeoutPrevote(height, round) {
+    if (height = h_p && round = round_p && step_p = prevote) {
+        precommit_extension ← ABCI.ExtendVote(h_p, round_p, nil)
```

Comment:
Could you add an example of `precommit_extension` in a comment?

Also, why would one need to extend the vote on `nil`? Is it a form of piggy-backing?

Reply:
Sure, I'll add an example of `precommit_extension`!

You have a good point, I don't actually foresee a reason to extend a `nil` vote (since that precommit won't make it into a block). 

That would only make sense in a consensus algorithm in which nil precommits made in time still get included in the next block, which doesn't happen at the moment. The only reasons I can imagine for this seem pretty contrived, so I'll remove it